jax/experimental/array_serialization/serialization.py

```diff
@@ -184,6 +185,20 @@ async def release_bytes(self, requested_bytes):
       self._cv.notify_all()
 
 
+async def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
+  data = shard.data
+  has_pinned_host = any(
+      m.kind == "pinned_host" for m in shard.device.addressable_memories())
+  if config.enable_memories.value and has_pinned_host:
```

Comment:
Why are we only adding asyncio.sleep(0) in the has_pinned_host case, wouldn't we want to address the issue even when memories are not enabled?

Also, I was doing some unit testing, and even without asyncio.sleep(0), it seems that the copy futures are still parallelized. I'm trying to understand how the issue with sequential copy futures arises, and whether it happens in some specific cases, or all cases.

Reply:
> Why are we only adding asyncio.sleep(0) in the has_pinned_host case, wouldn't we want to address the issue even when memories are not enabled?
You're right, there's not particular reason. I mostly wanted to leave the existing non-pinned-host copy path untouched in this PR. I pushed a newer revision that also achieves overlap in said case, by explicitly triggering the async copy.

Profiles of the non-host-pinned path before & after, also yielding a 2-3x improvement in my [self-contained example](https://github.com/google/jax/pull/22114#issuecomment-2196839593):
<img width="1452" alt="image" src="https://github.com/google/jax/assets/157012087/59557e9c-7066-4509-b1ad-ebabbe04859a">
<img width="1452" alt="image" src="https://github.com/google/jax/assets/157012087/aa6f36d8-44a4-41e9-ac12-6943ec6116e2">

Reply:
> Also, I was doing some unit testing, and even without asyncio.sleep(0), it seems that the copy futures are still parallelized. I'm trying to understand how the issue with sequential copy futures arises, and whether it happens in some specific cases, or all cases.

The copy futures are parallelized just fine -- the issue prior to https://github.com/google/jax/pull/22114 is that the synchronous part of `t.write(...)` ([here](https://github.com/google/jax/blob/e4b606e38a18867c757c738dd16265ca03d2cf88/jax/experimental/array_serialization/serialization.py#L263)) blocks on the device-to-host copy (before any copy future has been created). Said PR just disentangles the device-to-host copy from `t.write(...)` and allows overlapping copies of a single array's shards (i.e. all those copies originating within the same call of `async_serialize`, but in different calls of `_write_array`).